src/ripple/rpc/handlers/DownloadShard.cpp

```diff
@@ -135,8 +159,13 @@ doDownloadShard(RPC::JsonContext& context)
                 std::to_string(index) + " exists or being acquired");
         }
     }
+
+    // Begin downloading.
     if (!handler->start())
+    {
+        handler->release();
```

Comment:
If downloads were already running when `doDownloadShard` was invoked,  [this](https://github.com/undertome/rippled/blob/63d2d7ffe977f9907ac5cab1b563ab01217de80e/src/ripple/rpc/handlers/DownloadShard.cpp#L62) code block should return a non-error result that indicates that a download is already in progress.

Reply:
However, there is a small edge case I just discovered wherein there are no preshards, but the SAH has not yet completely reset its state and should not yet accept a new download request. So, at [this](https://github.com/undertome/rippled/blob/63d2d7ffe977f9907ac5cab1b563ab01217de80e/src/ripple/rpc/handlers/DownloadShard.cpp#L141) point in the code I plan to add a check to confirm that processing has completed in full. This will involve modest modifications to the SAH interface and private data.

Reply:
That's a good thing to discover.  Here's a hand-wave thought that you may have already been considering.  While you're restructuring, pause and think about whether it's possible to get the behavior you want not by adding a check but as a consequence of the restructure.  It certainly may not be possible.  But it's always worth considering whether a slightly different structure will more naturally give you the behavior you want.

And, no, I'm not smart enough to have something in mind.  But you may be smart enough to find that insight.

For an example of a very smart person (Sean Parent) being able to see a restructure you may find this video interesting: https://www.youtube.com/watch?v=IzNtM038JuI. It's not directly applicable to your code, but I find it inspiring...